# problem statement 

### 1.IMPORT REQUIRED LIBRARYS



In [116]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os 
from sklearn.model_selection import train_test_split
import pickle

In [32]:
 pd.set_option('display.float_format', '{:.2f}'.format)

###  2.READING THE DATA 

In [33]:
def get_data():

    #Noida  city csv file:
    jaipur=pd.read_csv('jaipur.csv')
    
    #Noida  city csv file:
    bhopal=pd.read_csv('bhopal.csv')

    
    #Noida  city csv file:
    vadodara=pd.read_csv('vadodara.csv')

    
    #Noida  city csv file:
    surat=pd.read_csv('surat.csv')

    
    #Noida  city csv file:
    kanpur=pd.read_csv('kanpur.csv')

    
    #Noida  city csv file:
    Guwahati=pd.read_csv('Guwahati.csv')

     #Noida  city csv file:
    Faridabad=pd.read_csv('Faridabad.csv')

    


    

    #Noida  city csv file:
    Lucknow=pd.read_csv('Lucknow.csv')

    #indore  city csv file:
    indore=pd.read_csv('indore.csv')

    #Noida  city csv file:
    Noida=pd.read_csv('Noida.csv')

    #Ghaziyabad  city csv file:
    Ghaziyabad=pd.read_csv('Ghaziyabad.csv')

    # Ahmadabad city csv file:
    Ahmadabad=pd.read_csv('Ahmadabad.csv')
    

    # kolkata city csv file:
    kolkata=pd.read_csv('kolkata.csv')

    
    # Hydrabadcity csv file:
    hyderabad=pd.read_csv('hyderabad.csv')
    
    
    # gurgaoncity csv file:
    gurgaon=pd.read_csv('gurgaon.csv')

    # Delhi city csv file:
    Delhi=pd.read_csv('Delhi.csv')
    
    # Chennai city csv file:
    Chennai=pd.read_csv('Chennai.csv')

    # Banglore city csv file:
    Banglore=pd.read_csv('Banglore.csv')
    
    # pune city csv file:
    pune=pd.read_csv('pune.csv')
    
    # mumbai city csv file:
    mumbai=pd.read_csv('mumbai.csv')

   ### merge alll above city into one dataframe:
    return  pd.concat([mumbai,pune,Banglore,Chennai,Delhi,gurgaon,hyderabad,kolkata,Ahmadabad,Ghaziyabad,Noida,indore,Lucknow,jaipur,bhopal,vadodara,surat,kanpur,Guwahati,Faridabad],ignore_index=True)

    
    

In [34]:
data=get_data()

In [35]:
data

broker_consultancy    Rent  Builtup Area  appartment_type  \
0      Kasturi Developers\nBUILDER\n-  20,000          1132                2   
1      Kasturi Developers\nBUILDER\n-  20,000          1200                2   
2      Kasturi Developers\nBUILDER\n-  30,000          1700                3   
3              Seller\nVERIFIED OWNER  28,500           750                2   
4              Seller\nVERIFIED OWNER  50,000          1400                3   
...                               ...     ...           ...              ...   
82011          Seller\nVERIFIED OWNER  11,000          1100                2   
82012          Seller\nVERIFIED OWNER   9,000           900                2   
82013          Seller\nVERIFIED OWNER  22,000          2727                4   
82014          Seller\nVERIFIED OWNER  11,000           600                2   
82015          Seller\nVERIFIED OWNER  40,000          1900                3   

              Deposit    Bathrooms            Facing                 Address  \
0          No Deposit  2 bathrooms  NorthEast facing            Ulwe, Mumbai   
1          No Deposit  2 bathrooms  NorthEast facing            Ulwe, Mumbai   
2          No Deposit  3 bathrooms  NorthEast facing            Ulwe, Mumbai   
3          No Deposit  2 bathrooms               NaN  Kandivali West, Mumbai   
4          No Deposit  3 bathrooms       East facing         Chembur, Mumbai   
...               ...          ...               ...                     ...   
82011  11,000 deposit  2 bathrooms               NaN     Sector 2, Faridabad   
82012   9,000 deposit  2 bathrooms               NaN    Sector 91, Faridabad   
82013  22,000 deposit  4 bathrooms               NaN    Sector 76, Faridabad   
82014      No Deposit  2 bathrooms               NaN    Sector 88, Faridabad   
82015      No Deposit  3 bathrooms  NorthEast facing    Sector 79, Faridabad   

              Apartment Name          Status  
0                  EV Castle     Unfurnished  
1                        NaN     Unfurnished  
2                        NaN  Semi-Furnished  
3             Atul Blue Arch     Unfurnished  
4               Godrej Prime  Semi-Furnished  
...                      ...             ...  
82011                    NaN     Unfurnished  
82012    Gupta Surya Nagar 2  Semi-Furnished  
82013                    NaN  Semi-Furnished  
82014                    NaN  Semi-Furnished  
82015  Omaxe Royal Residency  Semi-Furnished  

[82016 rows x 10 columns]

In [36]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 82016 entries, 0 to 82015
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   broker_consultancy  82015 non-null  object
 1   Rent                82016 non-null  object
 2   Builtup Area        82016 non-null  int64 
 3   appartment_type     82016 non-null  int64 
 4   Deposit             82016 non-null  object
 5   Bathrooms           81672 non-null  object
 6   Facing              28451 non-null  object
 7   Address             82016 non-null  object
 8   Apartment Name      32734 non-null  object
 9   Status              82016 non-null  object
dtypes: int64(2), object(8)
memory usage: 6.3+ MB


-   The data set contains 82,016 rows and 10 features
-   column "Bathrooms ,Facing,Appartment_Name," have lots of  missing values
-   The data type of some features are not appropriate

### 3.PREMLINARY DATA ANALYSIS

#### 3.1Check for datatype


In [37]:
data.sample(150)

broker_consultancy    Rent  Builtup Area  \
28379            Seller\nVERIFIED OWNER  32,800           550   
45185  Mahadev Renting Expert\nAGENT\n-  32,500          1100   
76399            Seller\nVERIFIED OWNER  16,000          1000   
50432            Seller\nVERIFIED OWNER  12,000           700   
62986     Datrika Enterprises\nAGENT\n0  56,000          1885   
...                                 ...     ...           ...   
17798            Seller\nVERIFIED OWNER   11000           650   
69713            Seller\nVERIFIED OWNER  17,000          1200   
35943         Neon Properties\nAGENT\n0  65,000          1500   
13246     National Consultant\nAGENT\n-   8,000           584   
81203            Seller\nVERIFIED OWNER  10,000           700   

       appartment_type         Deposit    Bathrooms            Facing  \
28379                1      No Deposit  1 bathrooms       East facing   
45185                3      No Deposit  2 bathrooms               NaN   
76399                2      No Deposit  2 bathrooms       East facing   
50432                1      No Deposit  1 bathrooms       East facing   
62986                3      No Deposit  3 bathrooms               NaN   
...                ...             ...          ...               ...   
17798                1      No Deposit  1 bathrooms               NaN   
69713                3  34,000 deposit  3 bathrooms               NaN   
35943                2      No Deposit  2 bathrooms  NorthEast facing   
13246                1   2,000 deposit  1 bathrooms               NaN   
81203                2      No Deposit  2 bathrooms               NaN   

                       Address         Apartment Name          Status  
28379     Bellandur, Bangalore                    NaN       Furnished  
45185         Janakpuri, Delhi     DDA Flat Janakpuri  Semi-Furnished  
76399    Vaibhav Nagar, Indore                    NaN  Semi-Furnished  
50432       Sector 46, Gurgaon                    NaN     Unfurnished  
62986  Nanakramguda, Hyderabad     Pacifica Hillcrest  Semi-Furnished  
...                        ...                    ...             ...  
17798        Anand Nagar, Pune                    NaN  Semi-Furnished  
69713    Chandkheda, Ahmedabad                    NaN  Semi-Furnished  
35943        Neeti Bagh, Delhi                    NaN  Semi-Furnished  
13246   Dombivali East, Mumbai  Lodha Palava Downtown  Semi-Furnished  
81203      Jalukbari, Guwahati                    NaN     Unfurnished  

[150 rows x 10 columns]

In [38]:
data.dtypes

broker_consultancy    object
Rent                  object
Builtup Area           int64
appartment_type        int64
Deposit               object
Bathrooms             object
Facing                object
Address               object
Apartment Name        object
Status                object
dtype: object

#### 3.2Check For Duplicates

In [39]:
data.duplicated().sum()

68994

In [40]:
duplicated_rows=(
    data.loc[data.duplicated(keep=False)].sort_values(["Rent","Builtup Area","appartment_type","Bathrooms","Facing","Address","Apartment Name","Status"])
)

In [41]:
duplicated_rows.sample(150)

broker_consultancy    Rent  Builtup Area  appartment_type  \
46432    CK Home Solutions\nAGENT\n-  15,000           850                2   
73309         Seller\nVERIFIED OWNER   5,000           400                1   
61292                 Ravi\nAGENT\n-  24,500           550                1   
56121         Seller\nVERIFIED OWNER  30,000          2400                4   
78315         Seller\nVERIFIED OWNER   6,000           600                1   
...                              ...     ...           ...              ...   
12196                  Seller\nOWNER   5,500           720                2   
42280         Seller\nVERIFIED OWNER  20,000          1300                3   
16811  Melange Consultancy\nAGENT\n0   48000          1300                2   
81265   ZROPERTIES REALITY\nAGENT\n0  13,000           800                1   
48725         Seller\nVERIFIED OWNER  16,000          1020                1   

              Deposit    Bathrooms Facing                         Address  \
46432  15,000 deposit  2 bathrooms    NaN                 Shahdara, Delhi   
73309   5,000 deposit  1 bathrooms    NaN  Nai Basti Dundahera, Ghaziabad   
61292      No Deposit  1 bathrooms    NaN             Kondapur, Hyderabad   
56121      No Deposit  4 bathrooms    NaN             Nizampet, Hyderabad   
78315   6,000 deposit  1 bathrooms    NaN             Jankipuram, Lucknow   
...               ...          ...    ...                             ...   
12196      No Deposit  2 bathrooms    NaN           Badlapur East, Mumbai   
42280      No Deposit  2 bathrooms    NaN              laxmi nagar, Delhi   
16811      No Deposit  2 bathrooms    NaN                  Hadapsar, Pune   
81265      No Deposit  1 bathrooms    NaN             Chandmari, Guwahati   
48725      No Deposit          NaN    NaN            PALAM VIHAR, Gurgaon   

      Apartment Name          Status  
46432            NaN  Semi-Furnished  
73309            NaN  Semi-Furnished  
61292            NaN       Furnished  
56121            NaN     Unfurnished  
78315       Oro City  Semi-Furnished  
...              ...             ...  
12196            NaN  Semi-Furnished  
42280            NaN       Furnished  
16811    Marvel Arco       Furnished  
81265            NaN  Semi-Furnished  
48725            NaN  Semi-Furnished  

[150 rows x 10 columns]

In [42]:
duplicated_rows.shape

(71650, 10)

In [43]:
82016-68994

13022

#### 3.3 Observations

- The type of `Rent`, `Deposit`, `bathroom`should be changed to numeric type.
- There're 6700 duplicates. These should be removed.

### 4.DETAILED ANALYSIS

#### broker_consultancy


In [44]:
data.broker_consultancy

0        Kasturi Developers\nBUILDER\n-
1        Kasturi Developers\nBUILDER\n-
2        Kasturi Developers\nBUILDER\n-
3                Seller\nVERIFIED OWNER
4                Seller\nVERIFIED OWNER
                      ...              
82011            Seller\nVERIFIED OWNER
82012            Seller\nVERIFIED OWNER
82013            Seller\nVERIFIED OWNER
82014            Seller\nVERIFIED OWNER
82015            Seller\nVERIFIED OWNER
Name: broker_consultancy, Length: 82016, dtype: object

In [45]:
data.broker_consultancy.dtypes

dtype('O')

In [46]:
data.broker_consultancy.isna().sum()

1

In [47]:
data.broker_consultancy.unique()

array(['Kasturi Developers\nBUILDER\n-', 'Seller\nVERIFIED OWNER',
       'CHERRISH ESTATE CONSULTANTS\nAGENT\n-', ...,
       'ROYALTY SOLUTION SERVICES\nAGENT\n0',
       'NARAYAN REALTORS GUWAHATI\nAGENT\n0', 'DEBOJIT\nAGENT\n0'],
      dtype=object)

-   The data type of broker_consultancy is valid
-   broker_consultancy contain one null value
-   There is no used of this feature in model_trainng
-   we are going drop this in data_cleaning operation




#### Rent

In [48]:
data.Rent

0        20,000
1        20,000
2        30,000
3        28,500
4        50,000
          ...  
82011    11,000
82012     9,000
82013    22,000
82014    11,000
82015    40,000
Name: Rent, Length: 82016, dtype: object

In [49]:
(
    data
    .Rent
    .loc[lambda ser: ser.str.contains("[^0-9,]")]
)

23       1.65 L
26        2.8 L
27        1.5 L
30        1.3 L
36        1.4 L
          ...  
81255       1 L
81275       1 L
81295       1 L
81315       1 L
81335       1 L
Name: Rent, Length: 4822, dtype: object

In [50]:
data.Rent.isna().sum()

0

In [51]:
data.Rent.dtypes

dtype('O')

In [52]:
(
    data
    .Rent
    .loc[lambda ser: ser.str.contains("[^0-9,]")]
     .str.replace("L","")
     .astype(float)
      .mul(100000)
      .astype(int)
      
    
    
)

23       165000
26       280000
27       150000
30       130000
36       140000
          ...  
81255    100000
81275    100000
81295    100000
81315    100000
81335    100000
Name: Rent, Length: 4822, dtype: int64

-   The data type of Rent  is not valid
-   In rent column data contain alphanumeric values like 1.5L,3.5L values
-   we need to convert this into the numeric value
-   This feature contain zero null value



#### Builtup_Area

-   The data type of Rent  is  valid
-   Does not contains any null value


#### Appartment_type

In [53]:
data.appartment_type.dtypes

dtype('int64')

In [54]:
data.appartment_type.isna().sum()

0

In [55]:
data.appartment_type.value_counts()

appartment_type
2    29133
3    25329
1    20777
4     5453
5     1164
6      125
7       32
8        2
9        1
Name: count, dtype: int64

-   The data type of Rent  is  valid
-   Does not contains any null value
-   we are going to remove the apparttment type which has ocuurance  less than 150 

#### Deposit

In [56]:
data.Deposit.unique()


array(['No Deposit', '4 L deposit', '1.25 L deposit', '1 L deposit',
       '60,000 deposit', '70,000 deposit', '3 L deposit', '3.5 L deposit',
       '1.5 L deposit', '3.3 L deposit', '2 L deposit', '6 L deposit',
       '9.75 L deposit', '10 L deposit', '9 L deposit', '80,000 deposit',
       '42,000 deposit', '75,000 deposit', '5 L deposit',
       '40,000 deposit', '2.5 L deposit', '5.4 L deposit', '20 L deposit',
       '15 L deposit', '18 L deposit', '8.5 L deposit', '2.1 L deposit',
       '1.74 L deposit', '1.35 L deposit', '28,000 deposit',
       '1.67 L deposit', '1.2 L deposit', '1.16 L deposit',
       '9.9 L deposit', '3.6 L deposit', '6.2 L deposit', '2.2 L deposit',
       '4.8 L deposit', '1.6 L deposit', '2.6 L deposit', '1.7 L deposit',
       '1.3 L deposit', '81,000 deposit', '84,000 deposit',
       '69,000 deposit', '1.08 L deposit', '1.02 L deposit',
       '1.56 L deposit', '54,000 deposit', '1.05 L deposit',
       '50,000 deposit', '7,000 deposit', '30,000 de

In [57]:
(
    data
    .Deposit
    .loc[lambda ser: ser.str.contains("[^No Deposit]")]
     .count()
    
)

12653

-   The data type of Deposit   is not  valid
-   This feature have lots of value as "no_Deposit"
-   because of that we are going to drop this feature

#### Bathrooms	

In [58]:
data.Bathrooms	

0        2 bathrooms
1        2 bathrooms
2        3 bathrooms
3        2 bathrooms
4        3 bathrooms
            ...     
82011    2 bathrooms
82012    2 bathrooms
82013    4 bathrooms
82014    2 bathrooms
82015    3 bathrooms
Name: Bathrooms, Length: 82016, dtype: object

In [59]:
data.Bathrooms.isna().sum()

344

In [60]:
(
    data
    .Bathrooms
     .str.strip()
     .str.get(0)
      
     
     
    
)

0        2
1        2
2        3
3        2
4        3
        ..
82011    2
82012    2
82013    4
82014    2
82015    3
Name: Bathrooms, Length: 82016, dtype: object

-   The data type of Bathrooms is not  valid
-   it should be numeric
-   this feature contains 344 null value
-   in datacleaning operation the above problems are handle

#### Facing


-   This feature have lots of null value
-   because of that we are going to drop this feature

#### Address

In [61]:
data.Address

0                  Ulwe, Mumbai
1                  Ulwe, Mumbai
2                  Ulwe, Mumbai
3        Kandivali West, Mumbai
4               Chembur, Mumbai
                  ...          
82011       Sector 2, Faridabad
82012      Sector 91, Faridabad
82013      Sector 76, Faridabad
82014      Sector 88, Faridabad
82015      Sector 79, Faridabad
Name: Address, Length: 82016, dtype: object

In [62]:
data.Address.isna().sum()

0

In [63]:
data['Address'].str.split(',').apply(lambda x:x[1]).unique()

array([' Mumbai', '', 'Thane west', ' Pune', ' Bangalore', ' Chennai',
       ' Delhi', ' Gurgaon', ' Hyderabad', ' Kolkata', ' Ahmedabad',
       ' Ghaziabad', ' Noida', ' Indore', ' Lucknow', ' Jaipur',
       ' Bhopal', ' Vadodara', ' Surat', ' Kanpur', ' Guwahati',
       ' Faridabad'], dtype=object)

In [64]:
len(data['Address'].str.split(',').apply(lambda x:x[1]).unique())

22

In [65]:
len(data['Address'].str.split(',').apply(lambda x:x[0]).unique())

626


-   This  Address feature have loction and city 
-   we have requirement of city name
-   because of that we have to seprate out the locality and city from address
-   add new feature as city and drop the feature address from dataframe
-   This feature does not contains missing values 

#### Apartment Name 

-   This feature have lots of null value
-   because of that we are going to drop this feature

#### Status

In [66]:
data.Status

0           Unfurnished
1           Unfurnished
2        Semi-Furnished
3           Unfurnished
4        Semi-Furnished
              ...      
82011       Unfurnished
82012    Semi-Furnished
82013    Semi-Furnished
82014    Semi-Furnished
82015    Semi-Furnished
Name: Status, Length: 82016, dtype: object

In [67]:
data.Status.isna().sum()

0

In [68]:
data.Status.unique()

array(['Unfurnished', 'Semi-Furnished', 'Furnished'], dtype=object)

-   This Feature has valid datatype
-   Does not contains any null value
-   not required any cleaning opertion

### 5.DATA CLEANING OPERATION


In [69]:
 def convert_rent(rent):
        if pd.isna(rent):  # Check for NaN values
            return np.nan
        if "L" in rent:
            return int(float(rent.replace('L', '').replace(',', '').strip()) * 100000)
        else:
            return int(float(rent.replace(',', '').strip()))

In [70]:
def clean_data(df):
    # Drop the 'broker_consultancy' column and rename columns to lowercase
    df_cleaned = (
        df
         
        .drop(columns="broker_consultancy")  # Drop the 'broker_consultancy','Deposit' column
        .drop(columns="Deposit")
        .drop(columns="Facing")
        .drop(columns="Apartment Name")
        #.drop_duplicates()
        
        .rename(columns=str.lower)            # Rename columns to lowercase
        .assign
        ( bathrooms=lambda df_:
         (
            df_
            
         .bathrooms
         .str.strip()
          .str.get(0)   
             
         )
            
        )
   
    )

    

    

    # Apply the convert_rent function to the 'rent' column
    df_cleaned['rent'] = df_cleaned['rent'].apply(convert_rent)

    if 'bathrooms' in df_cleaned.columns:
        mode_bathrooms = df_cleaned['bathrooms'].mode()[0]  # Calculate mode
        df_cleaned['bathrooms'] = df_cleaned['bathrooms'].fillna(mode_bathrooms) 

    # Convert 'bathrooms' column to integer type
    df_cleaned['bathrooms'] = df_cleaned['bathrooms'].astype(int)

    df_cleaned['city']=df_cleaned['address'].str.split(',').apply(lambda x:x[1])
    df_cleaned = df_cleaned.drop(columns="address")
    
    # Rename columns by removing or replacing substrings
    df_cleaned.columns = df_cleaned.columns.str.replace('area', '')
    df_cleaned.columns = df_cleaned.columns.str.replace('builtup', 'size')
    df_cleaned.columns = df_cleaned.columns.str.replace('appartment_type', 'type')
    df_cleaned.columns = df_cleaned.columns.str.replace('status', 'furnished_type')

    if 'type' in df_cleaned.columns:
        value_counts = df_cleaned['type'].value_counts()
        types_to_keep = value_counts[value_counts >= 150].index
        df_cleaned = df_cleaned[df_cleaned['type'].isin(types_to_keep)]

    df_cleaned['city'] =  df_cleaned['city'].str.strip()
    df_cleaned['city'] = df_cleaned['city'].replace('Thane west', 'Mumbai')
    df_cleaned['city'] = df_cleaned['city'].replace('','Mumbai')
    
    

    
     
    

    


    


    
    




    

    return df_cleaned
    


In [71]:
cleandata=clean_data(data)


In [72]:
cleandata.info()

<class 'pandas.core.frame.DataFrame'>
Index: 81856 entries, 0 to 82015
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   rent            81856 non-null  int64 
 1   size            81856 non-null  int64 
 2   type            81856 non-null  int64 
 3   bathrooms       81856 non-null  int64 
 4   furnished_type  81856 non-null  object
 5   city            81856 non-null  object
dtypes: int64(4), object(2)
memory usage: 4.4+ MB


In [73]:
cleandata.describe()

rent    size      type  bathrooms
count   81856.00 81856.00 81856.00   81856.00
mean    39035.96  1240.55     2.23       2.16
std     49955.71   750.24     0.95       0.97
min      1600.00   150.00     1.00       1.00
25%     15000.00   750.00     1.00       1.00
50%     25000.00  1100.00     2.00       2.00
75%     45000.00  1500.00     3.00       3.00
max   1800000.00  8326.00     5.00       7.00

In [74]:
cleandata.sample(150)

rent  size   type  bathrooms  furnished_type       city
52011   22000    700     1          1       Furnished    Gurgaon
9622    45000    806     2          2     Unfurnished     Mumbai
63006   56000   1885     3          3  Semi-Furnished  Hyderabad
50838   10000    300     1          1       Furnished    Gurgaon
10616   41500    651     1          2       Furnished     Mumbai
...       ...    ...   ...        ...             ...        ...
45409   28000   1100     3          2  Semi-Furnished      Delhi
464    140000   1250     2          2       Furnished     Mumbai
62916    9000    600     3          1       Furnished  Hyderabad
44099   11000    750     3          1  Semi-Furnished      Delhi
22773   24000   1250     2          2  Semi-Furnished       Pune

[150 rows x 6 columns]

In [75]:
len(cleandata.city.unique())

20

In [76]:
cleandata.city.value_counts()

city
Mumbai       16085
Delhi        14900
Hyderabad     8680
Bangalore     8246
Pune          7220
Gurgaon       6060
Ahmedabad     4760
Kolkata       3960
Noida         2820
Chennai       2033
Ghaziabad     1320
Indore        1320
Lucknow       1040
Faridabad      680
Guwahati       532
Surat          520
Jaipur         480
Vadodara       420
Kanpur         420
Bhopal         360
Name: count, dtype: int64

In [77]:
cleandata.type.value_counts()

type
2    29133
3    25329
1    20777
4     5453
5     1164
Name: count, dtype: int64

### 6.OUTLIERS DETECTIONS

In [78]:

def find_outliers(df, column_name):
    """
    Identifies outliers in the specified column of the DataFrame using the IQR method.
    """
    # Check if the column exists in the DataFrame
    if column_name not in df.columns:
        raise KeyError(f"Column '{column_name}' does not exist in the DataFrame.")
    
    # Calculate Q1 (25th percentile) and Q3 (75th percentile)
    Q1 = df[column_name].quantile(0.25)
    Q3 = df[column_name].quantile(0.75)
    # Calculate IQR
    IQR = Q3 - Q1
    # Determine outlier bounds
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    # Return boolean mask for outliers
    return (df[column_name] < lower_bound) | (df[column_name] > upper_bound)


### Rent

In [79]:

# Find outliers for 'rent'
rent_outliers = find_outliers(cleandata, 'rent')
print("Outliers for 'rent':")
print(rent_outliers)

Outliers for 'rent':
0        False
1        False
2        False
3        False
4        False
         ...  
82011    False
82012    False
82013    False
82014    False
82015    False
Name: rent, Length: 81856, dtype: bool


In [80]:
def drop_outliers(df, column_name):
    """
    Drops outliers from the specified column of the DataFrame and returns the cleaned DataFrame.
    """
    # Find outliers
    outlier_mask = find_outliers(df, column_name)
    
    # Ensure that outlier_mask is a boolean Series
    if not pd.api.types.is_bool_dtype(outlier_mask):
        raise TypeError("The outlier mask must be a boolean Series.")
    
    # Drop rows where the outlier mask is True
    df_cleaned = df[~outlier_mask]
    return df_cleaned

In [81]:
finaldata=drop_outliers(cleandata,'rent')

In [119]:
merge_data=drop_outliers(cleandata,'rent')


In [120]:
# Save the DataFrame to a CSV file
merge_data.to_csv('allcity.csv', index=False)


In [82]:
finaldata

rent  size   type  bathrooms  furnished_type       city
0      20000   1132     2          2     Unfurnished     Mumbai
1      20000   1200     2          2     Unfurnished     Mumbai
2      30000   1700     3          3  Semi-Furnished     Mumbai
3      28500    750     2          2     Unfurnished     Mumbai
4      50000   1400     3          3  Semi-Furnished     Mumbai
...      ...    ...   ...        ...             ...        ...
82011  11000   1100     2          2     Unfurnished  Faridabad
82012   9000    900     2          2  Semi-Furnished  Faridabad
82013  22000   2727     4          4  Semi-Furnished  Faridabad
82014  11000    600     2          2  Semi-Furnished  Faridabad
82015  40000   1900     3          3  Semi-Furnished  Faridabad

[76593 rows x 6 columns]

In [83]:
finaldata.describe()

rent    size      type  bathrooms
count 76593.00 76593.00 76593.00   76593.00
mean  29288.50  1147.61     2.15       2.07
std   19868.33   622.47     0.91       0.91
min    1600.00   150.00     1.00       1.00
25%   14500.00   700.00     1.00       1.00
50%   24000.00  1020.00     2.00       2.00
75%   40000.00  1350.00     3.00       2.00
max   90000.00  4200.00     5.00       6.00

In [118]:
# Save the DataFrame to a CSV file
finaldata.to_csv('output.csv', index=False)


### 7.Preprocessing Operations

In [84]:
finaldata

rent  size   type  bathrooms  furnished_type       city
0      20000   1132     2          2     Unfurnished     Mumbai
1      20000   1200     2          2     Unfurnished     Mumbai
2      30000   1700     3          3  Semi-Furnished     Mumbai
3      28500    750     2          2     Unfurnished     Mumbai
4      50000   1400     3          3  Semi-Furnished     Mumbai
...      ...    ...   ...        ...             ...        ...
82011  11000   1100     2          2     Unfurnished  Faridabad
82012   9000    900     2          2  Semi-Furnished  Faridabad
82013  22000   2727     4          4  Semi-Furnished  Faridabad
82014  11000    600     2          2  Semi-Furnished  Faridabad
82015  40000   1900     3          3  Semi-Furnished  Faridabad

[76593 rows x 6 columns]

-   This Dataset have city and furnished_type as categorical variable
-   we have to transform into the numeric value using Label_Encoder

In [85]:
from sklearn.preprocessing import LabelEncoder
encoder=LabelEncoder()

In [86]:
finaldata['city_encoded'] = encoder.fit_transform(finaldata['city'])
# Retrieve and print the mapping of cities to encoded values
city_mapping = dict(zip(encoder.classes_, encoder.transform(encoder.classes_)))
print("\nCity to Encoded Value Mapping:")
print(city_mapping)


City to Encoded Value Mapping:
{'Ahmedabad': 0, 'Bangalore': 1, 'Bhopal': 2, 'Chennai': 3, 'Delhi': 4, 'Faridabad': 5, 'Ghaziabad': 6, 'Gurgaon': 7, 'Guwahati': 8, 'Hyderabad': 9, 'Indore': 10, 'Jaipur': 11, 'Kanpur': 12, 'Kolkata': 13, 'Lucknow': 14, 'Mumbai': 15, 'Noida': 16, 'Pune': 17, 'Surat': 18, 'Vadodara': 19}


/tmp/ipykernel_4252/431990322.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  finaldata['city_encoded'] = encoder.fit_transform(finaldata['city'])


In [87]:
finaldata

rent  size   type  bathrooms  furnished_type       city  city_encoded
0      20000   1132     2          2     Unfurnished     Mumbai            15
1      20000   1200     2          2     Unfurnished     Mumbai            15
2      30000   1700     3          3  Semi-Furnished     Mumbai            15
3      28500    750     2          2     Unfurnished     Mumbai            15
4      50000   1400     3          3  Semi-Furnished     Mumbai            15
...      ...    ...   ...        ...             ...        ...           ...
82011  11000   1100     2          2     Unfurnished  Faridabad             5
82012   9000    900     2          2  Semi-Furnished  Faridabad             5
82013  22000   2727     4          4  Semi-Furnished  Faridabad             5
82014  11000    600     2          2  Semi-Furnished  Faridabad             5
82015  40000   1900     3          3  Semi-Furnished  Faridabad             5

[76593 rows x 7 columns]

In [88]:
finaldata['furnished_type'] = encoder.fit_transform(finaldata['furnished_type'])

/tmp/ipykernel_4252/127785236.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  finaldata['furnished_type'] = encoder.fit_transform(finaldata['furnished_type'])


In [89]:
finaldata

rent  size   type  bathrooms  furnished_type       city  city_encoded
0      20000   1132     2          2               2     Mumbai            15
1      20000   1200     2          2               2     Mumbai            15
2      30000   1700     3          3               1     Mumbai            15
3      28500    750     2          2               2     Mumbai            15
4      50000   1400     3          3               1     Mumbai            15
...      ...    ...   ...        ...             ...        ...           ...
82011  11000   1100     2          2               2  Faridabad             5
82012   9000    900     2          2               1  Faridabad             5
82013  22000   2727     4          4               1  Faridabad             5
82014  11000    600     2          2               1  Faridabad             5
82015  40000   1900     3          3               1  Faridabad             5

[76593 rows x 7 columns]

In [90]:
finaldata = finaldata.drop(columns='city')

In [91]:
finaldata

rent  size   type  bathrooms  furnished_type  city_encoded
0      20000   1132     2          2               2            15
1      20000   1200     2          2               2            15
2      30000   1700     3          3               1            15
3      28500    750     2          2               2            15
4      50000   1400     3          3               1            15
...      ...    ...   ...        ...             ...           ...
82011  11000   1100     2          2               2             5
82012   9000    900     2          2               1             5
82013  22000   2727     4          4               1             5
82014  11000    600     2          2               1             5
82015  40000   1900     3          3               1             5

[76593 rows x 6 columns]

In [92]:
finaldata.corr()

rent  size   type  bathrooms  furnished_type  city_encoded
rent            1.00   0.58  0.49       0.57           -0.12          0.15
size            0.58   1.00  0.75       0.81           -0.03         -0.02
type            0.49   0.75  1.00       0.83            0.00         -0.03
bathrooms       0.57   0.81  0.83       1.00            0.07          0.02
furnished_type -0.12  -0.03  0.00       0.07            1.00          0.19
city_encoded    0.15  -0.02 -0.03       0.02            0.19          1.00

In [93]:
model_data= finaldata.sample(frac=1, random_state=42).reset_index(drop=True)


In [94]:
model_data

rent  size   type  bathrooms  furnished_type  city_encoded
0      35000   1200     3          2               0             4
1      36000    785     2          2               1            15
2      50000   2050     3          3               1             9
3      24500    550     1          1               0             9
4      21500    600     1          1               0            15
...      ...    ...   ...        ...             ...           ...
76588  35000   1200     3          2               0             4
76589  45000    600     2          1               1            15
76590  56000   1885     3          3               1             9
76591  70000    923     2          2               1            15
76592  41000   1250     2          2               1            17

[76593 rows x 6 columns]

### 8.Split the data

In [95]:
X = model_data.drop(columns="rent")
y = model_data.rent.copy()

In [96]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [97]:
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(61274, 5) (61274,)
(15319, 5) (15319,)


### 9.Train The model & model selection


In [98]:
from sklearn.linear_model import LinearRegression

model_lr = LinearRegression()
model_lr.fit(X_train, y_train)

LinearRegression()

In [99]:
from sklearn.metrics import r2_score
r2 = r2_score(y_test, model_lr.predict(X_test))
r2

0.4110581382350328

In [100]:
# get the predicted data
y_pred = model_lr.predict(X_test)

# getting the true values
y_true = y_test

In [101]:
from sklearn.metrics import mean_absolute_error

# mean absolute error (MAE)
mae = mean_absolute_error(y_true, y_pred)
mae

11244.873757844676

In [102]:
from sklearn.metrics import mean_squared_error

# mean squared error (MSE)
mse = mean_squared_error(y_true, y_pred)
mse

231597455.08557147

In [103]:
from sklearn.metrics import r2_score

# R-squared score
r2 = r2_score(y_true, y_pred)
r2

0.4110581382350328

#### polynomial Regression 

In [104]:
from sklearn.preprocessing import PolynomialFeatures

# create a encoder object
polynomial_features = PolynomialFeatures(degree=5)

# preprocess the data to get the polynomials of X
x_polynomial = polynomial_features.fit_transform(X_train)
x_polynomial

array([[1.000000e+00, 9.500000e+02, 2.000000e+00, ..., 1.965200e+04,
        1.670420e+05, 1.419857e+06],
       [1.000000e+00, 5.000000e+02, 1.000000e+00, ..., 8.788000e+03,
        5.712200e+04, 3.712930e+05],
       [1.000000e+00, 7.500000e+02, 1.000000e+00, ..., 0.000000e+00,
        0.000000e+00, 2.430000e+02],
       ...,
       [1.000000e+00, 2.000000e+03, 1.000000e+00, ..., 1.000000e+00,
        1.000000e+00, 1.000000e+00],
       [1.000000e+00, 1.300000e+03, 2.000000e+00, ..., 7.290000e+02,
        6.561000e+03, 5.904900e+04],
       [1.000000e+00, 1.020000e+03, 3.000000e+00, ..., 0.000000e+00,
        0.000000e+00, 1.024000e+03]])

In [105]:
### create Linear Regression Model
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(x_polynomial, y_train)

LinearRegression()

In [106]:
# preprocess the data to get the polynomials of X
x_polynomial_test = polynomial_features.fit_transform(X_test)
x_polynomial_test

array([[1.000000e+00, 1.300000e+03, 3.000000e+00, ..., 4.913000e+03,
        8.352100e+04, 1.419857e+06],
       [1.000000e+00, 1.100000e+03, 3.000000e+00, ..., 8.788000e+03,
        5.712200e+04, 3.712930e+05],
       [1.000000e+00, 1.300000e+03, 2.000000e+00, ..., 7.290000e+02,
        6.561000e+03, 5.904900e+04],
       ...,
       [1.000000e+00, 1.020000e+03, 3.000000e+00, ..., 0.000000e+00,
        0.000000e+00, 1.024000e+03],
       [1.000000e+00, 2.800000e+03, 4.000000e+00, ..., 6.400000e+01,
        2.560000e+02, 1.024000e+03],
       [1.000000e+00, 6.000000e+02, 2.000000e+00, ..., 1.000000e+00,
        1.000000e+00, 1.000000e+00]])

In [107]:
from sklearn.metrics import r2_score

r2 = r2_score(y_test, model.predict(x_polynomial_test))
r2

0.3978434991045242

In [108]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import AdaBoostRegressor, RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
from xgboost import XGBRegressor
from catboost import CatBoostRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import SVR
from sklearn.naive_bayes import GaussianNB
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline


In [111]:
# Initialize models
models = {
    'Linear Regression': LinearRegression(),
    'Decision Tree': DecisionTreeRegressor(random_state=42),
    'XGBoost': XGBRegressor(random_state=42),
    'CatBoost': CatBoostRegressor(silent=True, random_state=42),
    'AdaBoost': AdaBoostRegressor(random_state=42),
    'Random Forest': RandomForestRegressor(random_state=42),
    
}

# Train and evaluate models
results = {}

for name, model in models.items():
    # Train the model
    model.fit(X_train, y_train)
    
    # Make predictions
    y_pred = model.predict(X_test)
    
    # Evaluate the model
    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
    mae = mean_absolute_error(y_test, y_pred)
    
    # Store the results
    results[name] = {
        'Mean Squared Error': mse,
        'R-squared': r2,
        'Mean Absolute Error': mae
    }

# Print the evaluation results
print("Model Evaluation Report:")
for name, metrics in results.items():
    print(f"\n{name}:")
    for metric, value in metrics.items():
        print(f"{metric}: {value:.4f}")

Model Evaluation Report:

Linear Regression:
Mean Squared Error: 231597455.0856
R-squared: 0.4111
Mean Absolute Error: 11244.8738

Decision Tree:
Mean Squared Error: 59034029.1247
R-squared: 0.8499
Mean Absolute Error: 2722.4265

XGBoost:
Mean Squared Error: 54843772.6932
R-squared: 0.8605
Mean Absolute Error: 2919.0184

CatBoost:
Mean Squared Error: 54321926.4595
R-squared: 0.8619
Mean Absolute Error: 3213.4955

AdaBoost:
Mean Squared Error: 186493165.1095
R-squared: 0.5258
Mean Absolute Error: 11047.8221

Random Forest:
Mean Squared Error: 56369835.1415
R-squared: 0.8567
Mean Absolute Error: 2715.0424


- CatBoost would be the best choice due to its lowest MSE and highest R-squared
- CatBoost stands out as the top performer

#### Final model for prediction

In [110]:
from catboost import CatBoostRegressor, Pool
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [115]:
model = CatBoostRegressor(
    iterations=500,          # Number of boosting iterations
    depth=6,                 # Depth of each tree
    learning_rate=0.1,       # Learning rate
    l2_leaf_reg=3,           # L2 regularization coefficient
    loss_function='RMSE',    # Loss function
    cat_features=[]          # Specify categorical features if any
)

# Fit the model
model.fit(X_train, y_train, eval_set=(X_test, y_test), verbose=100)

# Make predictions
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
print(f"Mean Squared Error: {mse}")
print(f"R2_Score: {r2}")
print(f"Mean Absolute Error: {mae}")
# Feature importances
feature_importances = model.get_feature_importance()
importance_df = pd.DataFrame({
    'Feature': X_train.columns,
    'Importance': feature_importances
}).sort_values(by='Importance', ascending=False)

print("\nFeature Importances:")
print(importance_df)

0:	learn: 18821.2537056	test: 18786.7581017	best: 18786.7581017 (0)	total: 17.3ms	remaining: 8.63s
100:	learn: 8155.2763847	test: 8311.7710899	best: 8311.7710899 (100)	total: 1.42s	remaining: 5.61s
200:	learn: 7601.6642962	test: 7806.8412529	best: 7806.8412529 (200)	total: 2.7s	remaining: 4.02s
300:	learn: 7368.5794032	test: 7614.3880755	best: 7614.3880755 (300)	total: 3.94s	remaining: 2.6s
400:	learn: 7240.2599860	test: 7518.6561099	best: 7518.5235391 (398)	total: 5.25s	remaining: 1.3s
499:	learn: 7138.4190574	test: 7445.2499361	best: 7445.2499361 (499)	total: 6.51s	remaining: 0us

bestTest = 7445.249936
bestIteration = 499

Mean Squared Error: 55431746.61088937
R2_Score: 0.859039573479601
Mean Absolute Error: 3448.008370066323

Feature Importances:
          Feature  Importance
4    city_encoded       42.14
0           size        32.10
3  furnished_type       11.03
2       bathrooms        8.17
1            type        6.56


In [117]:
# Save the model using pickle
with open('catboost_model.pkl', 'wb') as f:
    pickle.dump(model, f)

print("Model saved to catboost_model.pkl")

Model saved to catboost_model.pkl


In [121]:
import pandas as pd

# Example input data

# Encoded city (e.g., 3 for New York, 1 for Los Angeles)


# Convert to DataFrame
input_df = pd.DataFrame(input_data)


In [122]:
# Make predictions on the manually inputted data
predictions = model.predict(input_df)

# Output the predictions
print("Predictions:", predictions)


CatBoostError: /src/catboost/catboost/libs/data/model_dataset_compatibility.cpp:81: At position 0 should be feature with name size  (found size).

CatBoostError: /src/catboost/catboost/libs/data/model_dataset_compatibility.cpp:81: At position 0 should be feature with name size  (found size).


In [124]:
def test(a,b,c,d,e):
        
    input_data=[[a, b, c,d,e]]
    # Make predictions on the manually inputted data
    predictions = model.predict(input_data)
    
    # Output the predictions
    print("Predictions:", predictions)
    
    ## a=size
    ## b=type
    # c=noofbathrom
    # d=typeoffurnished
    # e=city

In [135]:
test(900,2,1,2,15)

Predictions: [36622.6376238]
